### PR TITLE
feat: migrate to DuckDB and modernize stack (Python 3.13, UV, Ruff)

### PR DESCRIPTION
**Pull request by ZdenekSrotyr**
# Migrate to DuckDB with Python 3.13 and modern tooling

## Summary

Replaces Snowflake workspace with local DuckDB processing for AWS Cost and Usage Reports. Modernizes the entire stack.

**Stack Upgrades:**
- Python 3.13 (from 3.10)
- UV package manager (replaces pip)
- Ruff linter (replaces flake8)
- `keboola.component` base class (replaces legacy `kbc` package)

**DuckDB Implementation:**
- Replaces Snowflake workspace with local DuckDB processing
- VIEW with lazy evaluation for streaming (no data materialization)
- Memory optimized: disk-backed DB, 1.5GB limit, single thread
- S3 support via httpfs extension

**Backward Compatibility:**
- Same normalization as v1.1.3 (manifest metadata)
- Column mapping preserved
- Output schema unchanged

**Performance:**
- Export time: ~1-2s for 21 files
- Peak memory: Constant (streaming) vs linear (old approach)
- No OOM errors on large reports

**Data flow:** `Files → VIEW (SQL) → COPY streaming → CSV`

---

### Updates since last revision

- **Dockerfile aligned with Keboola conventions** — Restructured to match the `component-acumatica` pattern: uses `UV_PROJECT_ENVIRONMENT="/usr/local/"` with `uv sync --all-groups --frozen` so packages install system-wide (no `.venv` needed). Removed `build-essential`, `PYTHONIOENCODING`, and `PYTHONPATH` env vars. Dependencies are now installed before copying source code for better Docker layer caching.
- **Dev dependencies moved to `[dependency-groups]`** — Migrated from `[project.optional-dependencies]` to `[dependency-groups]` in `pyproject.toml` (required by `uv sync`).
- **`uv.lock` regenerated** to reflect the new dependency group format.
- **Dockerfile-related commits squashed** into a single commit on top of the main squashed commit.

### Review history

All commits squashed into a single commit per @soustruh's suggestion, with one subsequent fix commit for the Dockerfile.

**Addressed review feedback:**
- @soustruh: imports moved to module level, `.gitignore` cleaned up, Dockerfile comments cleaned, `requires-python` tightened to `~=3.13.0`, Dockerfile changed to use `uv sync --all-groups --frozen` with `UV_PROJECT_ENVIRONMENT`
- @kudj: `__init__` simplified — removed custom `data_path_override` logic and manual debug handling already covered by `ComponentBase`

**Bug fixes discovered during production testing:**
1. `has_header=True` added to manifest — DuckDB exports CSV with header row
2. NULL fill for missing columns in incremental loads (schema shrinkage)
3. Dev dependencies separated to `[dependency-groups].dev`
4. Dead code removed (`_merge_case_variants`, `_normalize_column_name` from `duckdb_client.py`)
5. Column mapping conflict warnings downgraded to DEBUG
6. Manifest deduplication by billing period to prevent duplicate rows

---

## Review & Testing Checklist for Human

> ⚠️ **Risk: MEDIUM-HIGH** — Core data pipeline replacement with schema evolution edge cases. Verified on platform jobs but no automated integration test coverage for the new fixes.

- [ ] **CRITICAL: Revert CI component ID before merge** — `.github/workflows/push.yml` currently targets `kds-team.ex-aws-cost-and-usage-reports-test` instead of `kds-team.ex-aws-cost-and-usage-reports`. This **must** be changed back before merging.
- [ ] **Developer Portal update** — Remove Snowflake workspace requirement from component registration (no longer needed).
- [ ] **Verify column consistency end-to-end** — Run the production component config with DuckDB branch and compare output columns to v1.1.3 on the same config and date range.
- [ ] **Test incremental load with schema shrinkage** — Run a job that has state with N columns, then run again when source data has fewer columns. Confirm NULL-filled columns appear correctly and Storage accepts the load.
- [ ] **Verify `create_out_table_definition(schema=list[str])`** — Confirm that `keboola.component>=1.6.10` accepts a list of strings for the `schema` parameter and produces correct VARCHAR column definitions in the manifest.
- [x] **Verify Dockerfile `uv sync` works in CI** — ✅ CI passing (5/5 checks green). The new `uv sync --all-groups --frozen` with `UV_PROJECT_ENVIRONMENT="/usr/local/"` builds successfully and all dependencies are installed correctly.

**Recommended test plan:**
1. Reset state on test config, run full extraction with DuckDB branch
2. Compare output table columns to a v1.1.3 run on same config/date range
3. Run again (incremental) to verify state handling
4. Simulate schema shrinkage by modifying state to include extra columns, run again

### Notes

- NULL columns fix verified on job `1292911185` → subsequent run succeeded
- `has_header=True` fix verified on job `1292776725` (SUCCESS)
- Column deduplication edge case (two originals → same normalized name across manifests) identified but may still cause issues in rare scenarios
- PR approved by @kudj; @soustruh's feedback addressed
- CI passing with `uv sync` approach (5/5 checks green)
- Dockerfile now matches `component-acumatica` pattern for consistency across Keboola components

---

**Link to Devin run (original):** https://app.devin.ai/sessions/61bf4fb412d644e6804e65e7e8745f53  
**Link to Devin run (squash):** https://app.devin.ai/sessions/ccb8c2b5101b48faacbb4ee50e87664a  
**Requested by:** @ZdenekSrotyr

---

## Release Notes

**Justification, description**

Migration from Snowflake workspace to local DuckDB processing for AWS CUR extraction. Reduces infrastructure dependencies and improves performance.

**Plans for Customer Communication**

N/A — Internal implementation change, no user-facing API changes.

**Impact Analysis**

- Removes Snowflake workspace provisioning requirement
- Same output schema and normalization as v1.1.3
- Potential edge cases with case-insensitive column deduplication on accumulated state

**Deployment Plan**

1. Revert CI component ID to production
2. Merge and deploy
3. Update Developer Portal to remove workspace requirement

**Rollback Plan**

Revert to v1.1.3 tag if issues arise.

**Post-Release Support Plan**

Monitor Datadog logs for column mismatch errors on incremental loads.

### CODE REVIEW

File: component_config/configRowSchema.json
Comment (ZdenekSrotyr) [inline component_config/configRowSchema.json:1]: Tohlle není potřeba

File: component_config/configSchema.json
Comment (ZdenekSrotyr) [inline component_config/configSchema.json:1]: Config schema se meni proč? To by nemělo být potřeba

**Comment by ZdenekSrotyr** [comment]
Tested: 
https://connection.keboola.com/admin/projects/9231/branch/1274527/queue/1292958335
vs
https://connection.keboola.com/admin/projects/9231/branch/1274527/queue/1292911141

Devin validate a final tables https://app.devin.ai/sessions/cc0b6d05f3844b77aa5602df8cb1360d

File: src/component.py [hunk 1]
```diff
@@ -41,56 +42,44 @@
...
+        # Initialize ComponentBase
+        super().__init__(data_path_override=default_data_dir)
+
         # override debug from config
-        if self.cfg_params.get(KEY_DEBUG):
+        if self.configuration.parameters.get(KEY_DEBUG):
             debug = True
 
         if debug:
             logging.getLogger().setLevel(logging.DEBUG)
```
Comment (kudj) [inline src/component.py]: data path + debug is already solved by the ComponentBase

File: .github/workflows/push.yml [hunk 2]
```diff
@@ -122,7 +122,7 @@ jobs:
         run: |
           docker load --input /tmp/${{ env.KBC_DEVELOPERPORTAL_APP }}.tar
           docker image ls -a
-          docker run ${{ env.KBC_DEVELOPERPORTAL_APP }}:latest flake8 . --config=flake8.cfg
+          docker run ${{ env.KBC_DEVELOPERPORTAL_APP }}:latest ruff check .
```
Comment (kudj) [inline .github/workflows/push.yml:125]: yes!

**Review by kudj — DISMISSED** [review]
LGTM, nice performance in tests, just consider removing unnecessary code at the init that is already handled by ComponentBase.

**Review by kudj — DISMISSED** [review]
LGTM, thanks!

File: src/component.py [hunk 3]
```diff
@@ -234,14 +218,29 @@ def _download_and_unzip(self, key: str, local_path) -> str:
...
+        normalized_temp = self._kbc_normalize_header(original_cols)
+        normalized_temp = self._dedupe_header(normalized_temp)
+
+        # Map to canonical names from self.last_header (table columns)
+        canonical_map = {c.lower(): c for c in self.last_header}
+
+        # Filter to only columns that exist in both CSV and table
+        original_columns = []
+        normalized_columns = []
+        for orig, norm_temp in zip(original_cols, normalized_temp):
```
Comment (soustruh) [inline src/component.py:261]: I know this is probably vibed code, but I can still appreciate a pretty zip usage! 🤐 😍

File: src/duckdb_client.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+        """Compatibility method - connection is always open"""
+        pass
+
+    def close(self):
+        """Close the DuckDB connection and cleanup all temporary files"""
+        if self._connection:
+            self._connection.close()
+
+        # Clean up all DuckDB temporary files
+        import glob
```
Comment (soustruh) [inline src/duckdb_client.py]: Pretty please remove this import from the middle of the file 🙏

File: .gitignore [hunk 5]
```diff
@@ -11,4 +11,10 @@ test.py
...
 
-venv/
\ No newline at end of file
+# Virtual environments
+venv/
+.venv/
+
+# UV cache
+.uv/data_test/
+data_test/
```
Comment (soustruh) [inline .gitignore]: I don't think local testing directories have their place in the final gitignore, this way the gitignore files would just grow as every developer could add their own flavor of such folders… 🤷‍♂️ Maybe it's just my personal preference 🙃

File: Dockerfile [hunk 6]
```diff
@@ -1,17 +1,21 @@
...
 
-# install gcc to be able to build packages - e.g. required by regex, dateparser, also required for pandas
-RUN apt-get update && apt-get install -y build-essential
-
-RUN pip install --upgrade pip
-RUN pip install flake8
-
-RUN pip install -r /code/requirements.txt
+# Install gcc to be able to build packages - e.g. required by regex, dateparser
+RUN apt-get update && apt-get install -y build-essential && rm -rf /var/lib/apt/lists/*
```
Comment (soustruh) [inline Dockerfile]: No need to run slow apt operations at all probably?

File: Dockerfile [hunk 7]
```diff
@@ -1,17 +1,21 @@
...
 
 WORKDIR /code/
 
+# Copy pyproject.toml, README, source and tests
+COPY pyproject.toml README.md /code/
+COPY src/ /code/src/
+COPY tests/ /code/tests/
+
+# Install dependencies using UV (10-100x faster than pip)
```
Comment (soustruh) [inline Dockerfile]: LOL those comments… 😁

File: pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,42 @@
+[project]
+name = "component-aws-cost-and-usage-reports"
+version = "1.1.4"
+description = "AWS Cost and Usage Reports component for Keboola - DuckDB version"
+authors = [
+    {name = "Keboola", email = "support@keboola.com"}
+]
+readme = "README.md"
+license = {text = "MIT"}
+requires-python = ">=3.13"
```
Comment (soustruh) [inline pyproject.toml]: Please use the `~=3.13.0` specification, which means any Python in the 3.13 line, but not 3.14 (your definition means anythong lower than Python 4)

**Review by soustruh — CHANGES_REQUESTED** [review]
Just a couple of minor changes please 🙏

**Review by soustruh — DISMISSED** [review]
Thanks for all the changes! Due to the huge number of commits, I'd consider squashing this into a single commit with a nice and descriptive combined commit message? Leaving that ↑2U… 🙂

File: Dockerfile [hunk 9]
```diff
@@ -1,17 +1,16 @@
...
+# Install UV
+COPY --from=ghcr.io/astral-sh/uv:latest /uv /usr/local/bin/uv
 
 WORKDIR /code/
 
+COPY pyproject.toml README.md /code/
+COPY src/ /code/src/
+COPY tests/ /code/tests/
+
+RUN uv pip install --system --no-cache ".[dev]"
```
Comment (soustruh) [inline Dockerfile]: Why do we use uv pip install instead of
```
RUN uv sync --all-groups --frozen
```
like everywhere else? 🤔

**Review by soustruh — CHANGES_REQUESTED** [review]
I don't think having Dockerfile inconsistent with other projects isn't necessary? 🤷‍♂️

**Review by kudj — APPROVED** [review]
already approved the component refactor, now just minor fixes

**Review by soustruh — APPROVED** [review]
Nice, thanks for the changes ✅
